engine/src/main/java/org/hibernate/validator/internal/engine/cascading/ValueExtractors.java

```diff
@@ -64,13 +65,15 @@ public ValueExtractors(Iterable<ValueExtractor<?>> externalExtractors) {
 			for ( TypeVariable<?> cascadingValueTypeParameter : cascadingTypeParametersOfValueType ) {
 				TypeVariable<?> cascadingParameterBoundToExtractorType = bind( cascadingValueTypeParameter, bindingsForExtractorType );
 				if ( cascadingParameterBoundToExtractorType.equals( extractorDescriptor.extractedTypeParameter() ) ) {
-					cascadedValueExtractors.add( extractorDescriptor.getValueExtractor() );
+					// TODO implement selection of most specific extractor per requested type parameter
+					if ( !cascadedValueExtractors.containsKey( cascadingValueTypeParameter ) ) {
+						cascadedValueExtractors.put( cascadingValueTypeParameter, extractorDescriptor.getValueExtractor() );
+					}
 				}
 			}
 		}
 
-		// TODO implement selection of most specific extractor per requested type parameter
-		return cascadedValueExtractors.size() > 1 ? Collections.singletonList( cascadedValueExtractors.iterator().next() ) : cascadedValueExtractors;
+		return new ArrayList<>( cascadedValueExtractors.values() );
```

Comment:
That's what I thought and with this new implementation with an `HashMap`, you lose the order.